### PR TITLE
New class QgsJSONUtils with utilities for converting features to and from GeoJSON strings

### PR DESCRIPTION
**Pull request by nyalldawson**
Title says it all!

### CODE REVIEW

**Comment by m-kuhn** [comment]
:stuck_out_tongue_closed_eyes: e233b0b

nice stuff. did you checkout [qt5 json support](http://doc.qt.io/qt-5/json.html)? Do you think that can be valuable to us in the future?

**Comment by nyalldawson** [comment]
@m-kuhn yes, it sounds very nice! For 3.0 I'll add methods which return the Qt5 json objects too, but given that most of the time the desired end result will usually be a string anyway, the methods implemented here will still be useful/required.

File: src/core/qgsjsonutils.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    case QVariant::ULongLong:
+    case QVariant::Double:
+      return value.toString();
+
+    case QVariant::Bool:
+      return value.toBool() ? "true" : "false";
+
+    default:
+    case QVariant::String:
+      QString v = value.toString().replace( '"', "\\\"" )
```
Comment (rouault) [inline src/core/qgsjsonutils.cpp]: Looking at http://json.org/ , there are other characters ( \ \b \f \t / ) that would need to be \ escaped. See also https://github.com/json-c/json-c/blob/master/json_object.c#L109

Comment (nyalldawson) [inline src/core/qgsjsonutils.cpp]: Thanks @rouault! I'll fix this

**Comment by tomchadwin** [comment]
Literally just saw this:

 https://mailarchive.ietf.org/arch/msg/geojson/B5lKKb6mSf2YHAiJJXDfHs_uv-8

If adopted, do we import/export only GeoJSON v2?

**Comment by nyalldawson** [comment]
@tomchadwin hmmm... This current implementation ignores the crs and does not include any crs property. 

I'd be tempted to just say that QgsJSONUtils::featureToGeoJSON will ALWAYS expect geometries in WGS84. That way we'd already be compatible with GeoJSON v2. Possibly we could make featureToGeoJSON accept an optional coordinate transform (ie from the feature's native CRS to WGS84) and handle this conversion automatically if the transform is specified too.

**Comment by rldhont** [comment]
If we don't want to transform and add CRS information ?

**Comment by tomchadwin** [comment]
GeoJSON = EPSG:4326 has been an informal convention for a good while, so I think for export we are probably fine. But the link above suggests we should detect other CRS info and interpret as v1 if present, and therefore transform.

**Comment by nyalldawson** [comment]
I've added a bunch of new stuff:
- support for encoding lists and maps in json output
- I've moved the GeoJSON exporter to its own class (the interface was become to unwieldy for a single function)
- Extra properties can now be injected into the GeoJSON feature's properties

**Comment by tomchadwin** [comment]
> Extra properties can now be injected into the GeoJSON feature's properties

That could be a great feature. How about the converse - not including specified properties so that they are omitted from the GeoJSON?

**Comment by nyalldawson** [comment]
> That could be a great feature. How about the converse - not including specified properties so that they are omitted from the GeoJSON?

Ok - done. You could previously set a list of attributes to include manually, but I've added methods to set a list of attributes to specifically exclude too.

**Comment by nyalldawson** [comment]
@rldhont The GeoJSON export is now always reprojected to WGS84 - see 8f5757f and f0b0852. That should resolve http://hub.qgis.org/issues/14609

**Comment by nyalldawson** [comment]
Ok, last part I had planned for this just added - now you can choose to copy features from QGIS as GeoJSON FeatureCollections.

**Comment by nyalldawson** [comment]
Manually merged

**Comment by tomchadwin** [comment]
Splendid work, Nyall.

**Pull request closed without merging**